src/hotspot/share/prims/jvmtiImpl.cpp

```diff
@@ -185,6 +185,12 @@ void JvmtiBreakpoint::print_on(outputStream* out) const {
 //
 
 void VM_ChangeBreakpoints::doit() {
+ if (_bp->method() != Method::resolve_jmethod_id(_preservred_method)) {
+   // the jmethod_id's method was updated if class redefintion happened for this class
+   // after JvmtBreakpoint was created but before JVM_ChangeBreakpoints started
+   // all class breakpoints are cleared during redefinition so don't set/clear this breakpoint
```

Comment:
So basically one thread is trying to change this particular BP and it races with another thread that performs redefinition. If the redefinition thread wins then we are turning this current change into a no-op on the basis that the redefinition cleared all BPs anyway so we should not now set this one (if that was requested).

It is very unclear to me how the thread that requested the current change might respond to that request being ignored.

Please add some punctuation to the comment block as it is very hard to read at present. Thanks

Reply:
Thanks for feedback.

> It is very unclear to me how the thread that requested the current change might respond to that request being ignored.
> 
The SetBreakpoint returns 'JVMTI_ERROR_NONE' just like method set breakpoint and the breakpoint was removed later by class redfinition.

BTW The alternative would be to replace  _method with new version and set breakpoint like we first redefine classes and then set breakpoint. The both approaches looks valid for specification, I think.

> Please add some punctuation to the comment block as it is very hard to read at present. Thanks
will do this a little bit later.